File: csharp/examples/StepTrainV3.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Welvet;

namespace StepTrainV3
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("=== LOOM Stepping Neural Network v3: LSTM Middle Layer ===");
            Console.WriteLine("3-Layer Network: Dense -> LSTM -> Dense");
            Console.WriteLine();

            // 1. Define Network Architecture
            var config = new
            {
                batch_size = 1,
                grid_rows = 1,
                grid_cols = 3,
                layers_per_cell = 1,
                layers = new object[]
                {
                    new { type = "dense", input_height = 4, output_height = 8, activation = "relu" },
                    new { type = "lstm", input_size = 8, hidden_size = 12, seq_length = 1, activation = "tanh" },
                    new { type = "dense", input_height = 12, output_height = 3, activation = "softmax" }
                }
            };

            string jsonConfig = System.Text.Json.JsonSerializer.Serialize(config);
            Network.CreateFromJson(jsonConfig);

            // Initialize stepping state
            int inputSize = 4;
            using var state = new StepState(inputSize);

            // 2. Create Training Data (3 Classes)
            var trainingData = new[]
            {
                new { Input = new float[] { 0.1f, 0.2f, 0.1f, 0.3f }, Target = new float[] { 1.0f, 0.0f, 0.0f }, Label = "Low" },
                new { Input = new float[] { 0.2f, 0.1f, 0.3f, 0.2f }, Target = new float[] { 1.0f, 0.0f, 0.0f }, Label = "Low" },
                new { Input = new float[] { 0.8f, 0.9f, 0.8f, 0.7f }, Target = new float[] { 0.0f, 1.0f, 0.0f }, Label = "High" },
                new { Input = new float[] { 0.9f, 0.8f, 0.7f, 0.9f }, Target = new float[] { 0.0f, 1.0f, 0.0f }, Label = "High" },
                new { Input = new float[] { 0.1f, 0.9f, 0.1f, 0.9f }, Target = new float[] { 0.0f, 0.0f, 1.0f }, Label = "Mix" },
                new { Input = new float[] { 0.9f, 0.1f, 0.9f, 0.1f }, Target = new float[] { 0.0f, 0.0f, 1.0f }, Label = "Mix" }
            };

            // 3. Setup Continuous Training Loop
            int totalSteps = 100000;
            int targetDelay = 3;
            var targetQueue = new Queue<float[]>();

            float learningRate = 0.015f;
            float minLearningRate = 0.001f;
            float decayRate = 0.99995f;
            float gradientClipValue = 1.0f;

            Console.WriteLine($"Training for {totalSteps} steps (Max Speed)");
            Console.WriteLine($"Target Delay: {targetDelay} steps (accounts for LSTM internal state)");
            Console.WriteLine($"LR Decay: {decayRate:F4} per step (min {minLearningRate:F4})");
            Console.WriteLine($"Gradient Clipping: {gradientClipValue:F2}");
            Console.WriteLine();

            var startTime = DateTime.Now;
            int stepCount = 0;
            int currentSampleIdx = 0;
            var random = new Random();

            Console.WriteLine($"{"Step",-6} {"Input",-10} {"Output (ArgMax)",-25} {"Loss",-10}");
            Console.WriteLine("──────────────────────────────────────────────────────────");

            while (stepCount < totalSteps)
            {
                if (stepCount % 20 == 0)
                {
                    currentSampleIdx = random.Next(trainingData.Length);
                }

                var sample = trainingData[currentSampleIdx];

                // B. Set Input
                state.SetInput(sample.Input);

                // C. Step Forward
                state.StepForward();

                // D. Manage Target Queue
                targetQueue.Enqueue(sample.Target);

                if (targetQueue.Count >= targetDelay)
                {
                    var delayedTarget = targetQueue.Dequeue();
                    var output = state.GetOutput();

                    // F. Calculate Loss & Gradient
                    float loss = 0.0f;
                    var gradOutput = new float[output.Length];

                    for (int i = 0; i < output.Length; i++)
                    {
                        float p = output[i];
                        if (p < 1e-7f) p = 1e-7f;
                        if (p > 1.0f - 1e-7f) p = 1.0f - 1e-7f;

                        if (delayedTarget[i] > 0.5f)
                        {
                            loss -= (float)Math.Log(p);
                        }

                        gradOutput[i] = output[i] - delayedTarget[i];
                    }

                    // Apply gradient clipping
                    float gradNorm = 0.0f;
                    foreach (var g in gradOutput) gradNorm += g * g;
                    gradNorm = (float)Math.Sqrt(gradNorm);

                    if (gradNorm > gradientClipValue)
                    {
                        float scale = gradientClipValue / gradNorm;
                        for (int i = 0; i < gradOutput.Length; i++)
                        {
                            gradOutput[i] *= scale;
                        }
                    }

                    // G. Backward Pass
                    state.StepBackward(gradOutput);

                    // H. Update Weights
                    Network.ApplyGradients(learningRate);

                    // Decay Learning Rate
                    learningRate *= decayRate;
                    if (learningRate < minLearningRate) learningRate = minLearningRate;

                    // I. Logging
                    if (stepCount % 500 == 0)
                    {
                        int maxIdx = 0;
                        float maxVal = output[0];
                        for (int i = 1; i < output.Length; i++)
                        {
                            if (output[i] > maxVal)
                            {
                                maxVal = output[i];
                                maxIdx = i;
                            }
                        }

                        int tMaxIdx = 0;
                        for (int i = 1; i < delayedTarget.Length; i++)
                        {
                            if (delayedTarget[i] > 0.5f) tMaxIdx = i;
                        }

                        string mark = maxIdx == tMaxIdx ? "✓" : "✗";

                        Console.WriteLine($"{stepCount,-6} {sample.Label,-10} Class {maxIdx} ({maxVal:F2}) [{mark}] Exp: {tMaxIdx}  Loss: {loss:F4}  LR: {learningRate:F5}");
                    }
                }

                stepCount++;
            }

            var totalTime = DateTime.Now - startTime;
            Console.WriteLine();
            Console.WriteLine("=== Training Complete ===");
            Console.WriteLine($"Total Time: {totalTime.TotalSeconds:F3}s");
            Console.WriteLine($"Speed: {totalSteps / totalTime.TotalSeconds:F2} steps/sec");
            Console.WriteLine();

            // Final Evaluation
            Console.WriteLine("Evaluating on all samples (with settling time)...");
            int correct = 0;
            int settlingSteps = 10;

            foreach (var sample in trainingData)
            {
                state.SetInput(sample.Input);
                for (int i = 0; i < settlingSteps; i++)
                {
                    state.StepForward();
                }

                var output = state.GetOutput();
                int maxIdx = 0;
                float maxVal = output[0];
                for (int i = 1; i < output.Length; i++)
                {
                    if (output[i] > maxVal)
                    {
                        maxVal = output[i];
                        maxIdx = i;
                    }
                }

                int tMaxIdx = 0;
                for (int i = 1; i < sample.Target.Length; i++)
                {
                    if (sample.Target[i] > 0.5f) tMaxIdx = i;
                }

                string mark = "✗";
                if (maxIdx == tMaxIdx)
                {
                    correct++;
                    mark = "✓";
                }

                Console.WriteLine($"{mark} {sample.Label}: Pred {maxIdx} ({maxVal:F2}) Exp {tMaxIdx}");
            }

            Console.WriteLine($"Final Accuracy: {correct}/{trainingData.Length}");
        }
    }
}

File: python/examples/step_train_v3.py
#!/usr/bin/env python3
"""
Stepping Network Training Example (Python Version)
Replicates the exact logic and output format of examples/step_example/step_train_v3.go
"""

import json
import time
import random
import math
from welvet import (
    create_network_from_json,
    StepState,
    apply_gradients,
)

class TargetQueue:
    """Handles the delay between input and output in the stepping network."""
    def __init__(self, size):
        self.targets = []
        self.max_size = size

    def push(self, target):
        self.targets.append(target)

    def pop(self):
        if not self.targets:
            return None
        return self.targets.pop(0)

    def is_full(self):
        return len(self.targets) >= self.max_size

def main():
    print("=== LOOM Stepping Neural Network v3: LSTM Middle Layer ===")
    print("3-Layer Network: Dense -> LSTM -> Dense")
    print()

    # 1. Define Network Architecture
    config = {
        "batch_size": 1,
        "grid_rows": 1,
        "grid_cols": 3,
        "layers_per_cell": 1,
        "layers": [
            {
                "type": "dense",
                "input_height": 4,
                "output_height": 8,
                "activation": "relu"
            },
            {
                "type": "lstm",
                "input_size": 8,
                "hidden_size": 12,
                "seq_length": 1,
                "activation": "tanh"
            },
            {
                "type": "dense",
                "input_height": 12,
                "output_height": 3,
                "activation": "softmax"
            }
        ]
    }

    create_network_from_json(config)

    # Initialize stepping state
    input_size = 4
    state = StepState(input_size=input_size)

    # 2. Create Training Data (3 Classes)
    training_data = [
        # Class 0: Low values
        {"input": [0.1, 0.2, 0.1, 0.3], "target": [1.0, 0.0, 0.0], "label": "Low"},
        {"input": [0.2, 0.1, 0.3, 0.2], "target": [1.0, 0.0, 0.0], "label": "Low"},
        # Class 1: High values
        {"input": [0.8, 0.9, 0.8, 0.7], "target": [0.0, 1.0, 0.0], "label": "High"},
        {"input": [0.9, 0.8, 0.7, 0.9], "target": [0.0, 1.0, 0.0], "label": "High"},
        # Class 2: Mixed
        {"input": [0.1, 0.9, 0.1, 0.9], "target": [0.0, 0.0, 1.0], "label": "Mix"},
        {"input": [0.9, 0.1, 0.9, 0.1], "target": [0.0, 0.0, 1.0], "label": "Mix"},
    ]

    # 3. Setup Continuous Training Loop
    total_steps = 100000
    target_delay = 3
    target_queue = TargetQueue(target_delay)

    learning_rate = 0.015
    min_learning_rate = 0.001
    decay_rate = 0.99995
    gradient_clip_value = 1.0

    print(f"Training for {total_steps} steps (Max Speed)")
    print(f"Target Delay: {target_delay} steps (accounts for LSTM internal state)")
    print(f"LR Decay: {decay_rate:.4f} per step (min {min_learning_rate:.4f})")
    print(f"Gradient Clipping: {gradient_clip_value:.2f}")
    print()

    start_time = time.time()
    step_count = 0
    current_sample_idx = 0

    print(f"{'Step':<6} {'Input':<10} {'Output (ArgMax)':<25} {'Loss':<10}")
    print("──────────────────────────────────────────────────────────")

    while step_count < total_steps:
        # Rotate sample every 20 steps
        if step_count % 20 == 0:
            current_sample_idx = random.randint(0, len(training_data) - 1)
        
        sample = training_data[current_sample_idx]

        # B. Set Input
        state.set_input(sample["input"])

        # C. Step Forward
        state.step_forward()

        # D. Manage Target Queue
        target_queue.push(sample["target"])

        if target_queue.is_full():
            delayed_target = target_queue.pop()
            output = state.get_output()

            # F. Calculate Loss & Gradient
            loss = 0.0
            grad_output = [0.0] * len(output)

            for i in range(len(output)):
                p = output[i]
                # Clamp for numerical stability
                p = max(1e-7, min(1.0 - 1e-7, p))

                if delayed_target[i] > 0.5:
                    loss -= math.log(p)

                # Gradient for Softmax + CrossEntropy
                grad_output[i] = output[i] - delayed_target[i]

            # Apply gradient clipping
            grad_norm = sum(g * g for g in grad_output)
            grad_norm = math.sqrt(grad_norm)

            if grad_norm > gradient_clip_value:
                scale = gradient_clip_value / grad_norm
                for i in range(len(grad_output)):
                    grad_output[i] *= scale

            # G. Backward Pass
            state.step_backward(grad_output)

            # H. Update Weights
            apply_gradients(learning_rate)

            # Decay Learning Rate
            learning_rate *= decay_rate
            learning_rate = max(learning_rate, min_learning_rate)

            # I. Logging
            if step_count % 500 == 0:
                max_idx = output.index(max(output))
                max_val = output[max_idx]
                
                t_max_idx = delayed_target.index(max(delayed_target))
                
                mark = "✗"
                if max_idx == t_max_idx:
                    mark = "✓"

                print(f"{step_count:<6} {sample['label']:<10} Class {max_idx} ({max_val:.2f}) [{mark}] Exp: {t_max_idx}  Loss: {loss:.4f}  LR: {learning_rate:.5f}")

        step_count += 1

    total_time = time.time() - start_time
    print()
    print("=== Training Complete ===")
    print(f"Total Time: {total_time:.3f}s")
    print(f"Speed: {total_steps/total_time:.2f} steps/sec")
    print()

    # Final Evaluation
    print("Evaluating on all samples (with settling time)...")
    correct = 0
    settling_steps = 10

    for sample in training_data:
        state.set_input(sample["input"])
        # Settle
        for _ in range(settling_steps):
            state.step_forward()
            
        output = state.get_output()
        max_idx = output.index(max(output))
        max_val = output[max_idx]
        
        t_max_idx = sample["target"].index(max(sample["target"]))
        
        mark = "✗"
        if max_idx == t_max_idx:
            correct += 1
            mark = "✓"
            
        print(f"{mark} {sample['label']}: Pred {max_idx} ({max_val:.2f}) Exp {t_max_idx}")
        
    print(f"Final Accuracy: {correct}/{len(training_data)}")

if __name__ == "__main__":
    main()
